### ISSUE TITLE
ags3: font issues with 1271.Maniac Mansion Mania

### ISSUE THREAD

**Issue by rofl0r** [adventuregamestudio/ags#655]
1271.Maniac Mansion Mania - Episode 16 Return of the Meteor [2007 reloaded version].zip

https://0x0.st/zo8b.zip

immediately on the intro, and first room, there are sometimes some issues with the font:
screenshot attached
![1](https://0x0.st/zo8c.png)

version used: 3.4.3.1 on linux, allegro 4.4.3

**Comment by ivan-mogilko** [comment]
If it's a TTF font, that might be related to differences in freefont library versions (used in original AGS and currently used for Linux port).

Our own alfont fork which we use for Windows version includes compatible freefont sources (https://github.com/adventuregamestudio/lib-alfont). What I wonder, may that be possible to require using it on linux too? If that worked to fix these compatibility issues it could be the solution for ags3 branch.

Alternate solution is to do some hack, similar to this (may be related since it's also about outline):
https://github.com/adventuregamestudio/ags/blob/master/Common/font/ttffontrenderer.cpp#L102

**Comment by rofl0r** [comment]
thanks, very helpful. turned out that the game uses the ttf font "LucasFan-Font" (precisely the one requiring the "hack"), however the hack doesn't kick in because the condition `get_font_outline(fontNumber) >= 0` isn't true (the value is -1).

by forcing the hack to become active even with -1 nothing changed, and that's because the inner yellow text uses the ttf font whereas the black outline uses a wfn version of the same font, and that doesn't seem to play well together. i found this by trying to replace the lucasfan .ttf font copies with another ttf font after unpacking the game with [agstract](https://github.com/rofl0r/agsutils) and repacking it after the mod.

due to these circumstances, i would expect the issue happens in windows as well. maybe i'll find a properly working ttf/wfn pair that can be used to replace the ones used... a completely transparent wfn font could help too, i guess, that could allow to hack the game to use only the ttf font with an outline...

**Comment by rofl0r** [comment]
@ivan-mogilko why is the lib-alfont repo on 1.9.1 and the alfont source shipped  (and built) in Engine/libsrc 2.0.9 ?

**Comment by ivan-mogilko** [comment]
Originally we had prebuilt libraries on Windows, and IIRC it took time to find out what version of the library they were built from and reconstruct the changes made by Chris Jones to them. Alfont 2.0.9 was the last existing alfont library, I think, that's why JJS probably chose it as a base to apply these extra changes.

**Comment by rofl0r** [comment]
i think i can say with great confidence that the windows version uses the 2.0.9 code too, because the 1.9.1 code in the lib-alfont repo lacks `alfont_get_name()` which is used by the hack you pointed out.

meanwhile it seems my previous assessment that the outline was created by a wfn font is wrong, because i replaced all 3 wfn fonts with `transparent.wfn` generated by this python snippet:
```py
# create empty wfn font
def le(num):
  return ('%02x%02x'%(num % 256, num / 256)).decode('hex')

hdr = "WGT Font File  "
char = '08000800' + ('00'*8)
char = char.decode('hex')
chardata = char*128
off_tbl_off = len(hdr) + 2 + len(chardata)
off_tbl_off_bin_le = le(off_tbl_off)
off = len(hdr)+2
offs = ''
for i in range(128):
  offs = offs + le(off)
  off += 12 # width (2b) + height (2b) + chardata (8b)
wfn = hdr + off_tbl_off_bin_le + chardata + offs
open('transparent.wfn', 'w').write(wfn)
```
yet the outline is still visible.
this is really odd... is it possible the outline is saved as sprites ?

**Comment by rofl0r** [comment]
turns out that font `agsfnt2.ttf` called `Arcade` is used for the outline, whereas font `agsfnt1.ttf``Lucasfan...` is used for the yellow content.

meanwhile i managed to hack ags sources to include lib-alfont 1.9.1 and link against a static linked freetype 2.1.3 (found here: http://ftp.math.utah.edu/pub/freetype/ ) which is equal to the one in the lib-alfont repo (minus DOS line endings and some windows makefile changes in the ags repo). using that setup, everything looks precisely as it should.
maybe one could go on a hunt to find the latest freetype version that works with ags and possibly identify the breaking change by comparing it with the first version that breaks.

**Comment by rofl0r** [comment]
update: work with alfont 2.0.9 too. so it looks as if only the freetype version is critical.

**Comment by rofl0r** [comment]
fixed via #690
